ipn/ipnlocal/serve.go

```diff
@@ -37,6 +42,166 @@ type serveHTTPContext struct {
 	DestPort uint16
 }
 
+type serveListener struct {
+	ap     netip.AddrPort
+	ctx    context.Context    // valid while listener is desired
+	cancel context.CancelFunc // for ctx, to close listener
+	logf   logger.Logf
+
+	ln net.Listener     // non-nil if listener is active
+	bo *backoff.Backoff // for retrying failed Listen calls
+}
+
+func newServeListener(ctx context.Context, ap netip.AddrPort, logf logger.Logf) *serveListener {
+	ctx, cancel := context.WithCancel(ctx)
+	return &serveListener{
+		ap:     ap,
+		ctx:    ctx,
+		cancel: cancel,
+		logf:   logf,
+
+		bo: backoff.NewBackoff("serve-listener", logf, 30*time.Second),
+	}
+}
+
+func (s *serveListener) Close() error {
+	s.cancel()
+	return s.ln.Close()
```

Comment:
I think this s.ln access will be a data race.

make the ln field be a `syncs.AtomicValue[net.Listener]` instead?